### PR TITLE
Use input_num to remove array creation overhead

### PR DESCRIPTION
**Pull request by unnonouno**
To remove redundant array creation overhead, I used `input_num`. Merge https://github.com/cupy/cupy/pull/868 first.

### CODE REVIEW

**Comment by okuta** [comment]
jenkins, test this plase.

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by okuta** [comment]
Please check travis result.

**Comment by unnonouno** [comment]
#4002 has higher priority. It does not block this PR, but it causes conflict. Please check #4002 first.

**Comment by unnonouno** [comment]
Hum, it looks that when we want to use `cupy.fusion.*` which actually supports `numpy.ndarray` it cannot be applied for an environment without CUDA...

**Comment by okuta** [comment]
Thank you for your fixing

**Comment by okuta** [comment]
If fusion function supports `__setitem__`, we can use more simple syntax maybe.

**Comment by okuta** [comment]
If fusion supports `__setitem__`, we can use more simple syntax.

**Comment by unnonouno** [comment]
Is it difficult? I think it sounds great.

**Comment by okuta** [comment]
I cannot fix immediately.

File: chainer/functions/activation/lstm.py [hunk 1]
```diff
@@ -235,25 +236,34 @@ def lstm_grad_grad(
...
     dgc_do = gh * gsig_o * gtanh_c
-    go = ggc * dgc_do + ggo * gh * tanh_c * ggsig_o
+    xp.add(ggc * dgc_do, ggo * gh * tanh_c * ggsig_o, out=go)
     dgc_dc = gh * sig_o * ggtanh_c
-    gc_next = ggc * dgc_dc + ggo * gh * gtanh_c * gsig_o
-    ggh = ggc * sig_o * gtanh_c + ggo * tanh_c * gsig_o
+    xp.add(
+        ggc * dgc_dc, ggo * gh * gtanh_c * gsig_o, out=gc_next)
+    xp.add(ggc * sig_o * gtanh_c, ggo * tanh_c * gsig_o, out=ggh)
     return gc_prev, ga, gi, gf, go, gc_next, ggc, ggh
```
Comment (okuta) [inline chainer/functions/activation/lstm.py]: Is it necessary?

**Comment by okuta** [comment]
> BTW, I feel this fix looks complicated and hard to read. Is there any other method to write it in more beautiful syntax?

I implemented it. Cloud you use it?

File: chainer/functions/activation/lstm.py (hunk 1 above)
Comment (unnonouno) [reply chainer/functions/activation/lstm.py]: I found that when i remove `return`, a compile error occurs. It is a bug of fuse.

**Comment by unnonouno** [comment]
See https://github.com/cupy/cupy/pull/1020

**Comment by okuta** [comment]
cupy/cupy#1020 was merged.

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by okuta** [comment]
jenkins, test this please.

File: chainer/functions/activation/lstm.py [hunk 2]
```diff
@@ -214,9 +214,10 @@ def _cupy_sigmoid(x):
     return cuda.fusion.tanh(x * half) * half + half
 
 
-@cuda.fuse()
+@cuda.fuse(input_num=13)
```
Comment (okuta) [inline chainer/functions/activation/lstm.py]: Please remove `input_num=13` or return line.

**Comment by kmaehashi** [comment]
I set milestone to v5.0.0a1. 
If this fix is needed for v4, we need to add `to-be-backported` so that we can include this into v4 final release.

File: chainer/functions/activation/lstm.py (hunk 2 above)
Comment (okuta) [reply chainer/functions/activation/lstm.py]: @unnonouno Do you have time to fix this?

**Comment by unnonouno** [comment]
I removed `input_num`. Is it correct?

**Comment by okuta** [comment]
LGTM!

**Pull request merged**